File: benchmarks/fixtures/webapp_ts/src/services/payment/processor.ts
/** Payment processor with diamond-like inheritance. */
import { getLogger, generateRequestId } from '../../utils/helpers';
import { CacheableService } from '../cacheable';
import { Auditable } from '../base';
import { DatabaseConnection } from '../../database/connection';
import { PaymentQueries } from '../../database/queries';
import { EventDispatcher } from '../../events/dispatcher';
import { PaymentError, ValidationError, NotFoundError } from '../../errors';

const logger = getLogger("services.payment");

const SUPPORTED_CURRENCIES = ["USD", "EUR", "GBP", "JPY", "CAD"];

/** Payment processor with caching + audit. */
export class PaymentProcessor extends CacheableService implements Auditable {
    private events: EventDispatcher;
    private queries: PaymentQueries;
    private auditLog: Array<{ action: string; actor: string; resource: string; details: Record<string, unknown>; timestamp: number }> = [];

    constructor(db: DatabaseConnection, events: EventDispatcher) {
        super("payment_processor");
        this.events = events;
        this.queries = new PaymentQueries(db);
    }

    /** Process a payment. */
    async processPayment(userId: string, amount: number, currency: string, method: string = "card"): Promise<Record<string, unknown>> {
        this.requireInitialized();
        logger.info(`Processing payment: user=${userId}, amount=${amount} ${currency}`);
        this.validatePayment(amount, currency);
        const txnId = generateRequestId();
        const cacheKey = `payment:${userId}:${amount}:${currency}`;
        if (this.cacheGet(cacheKey)) {
            throw new PaymentError("Duplicate payment", txnId);
        }
        try {
            await this.queries.createPayment(userId, amount, currency, txnId);
            await this.queries.updateStatus(txnId, "completed");
        } catch (e) {
            throw new PaymentError(`Payment failed: ${e}`, txnId);
        }
        this.cacheSet(cacheKey, txnId, 300);
        this.recordAudit("payment.processed", userId, `payment:${txnId}`, { amount, currency, method });
        this.events.emit("payment.completed", { transactionId: txnId, userId, amount, currency });
        return { transactionId: txnId, status: "completed", amount, currency };
    }

    /** Refund a payment. */
    async refund(transactionId: string, reason: string = ""): Promise<Record<string, unknown>> {
        logger.info(`Refunding: ${transactionId}`);
        const payment = await this.queries.findByTransactionId(transactionId);
        if (!payment) throw new NotFoundError("Payment", transactionId);
        await this.queries.updateStatus(transactionId, "refunded");
        this.recordAudit("payment.refunded", "system", `payment:${transactionId}`, { reason });
        this.events.emit("payment.refunded", { transactionId, reason });
        return { transactionId, status: "refunded" };
    }

    /** Record an audit entry. */
    recordAudit(action: string, actor: string, resource: string, details?: Record<string, unknown>): void {
        this.auditLog.push({ action, actor, resource, details: details ?? {}, timestamp: Date.now() });
        logger.info(`Audit: ${actor} ${action} on ${resource}`);
    }

    /** Get audit trail. */
    getAuditTrail(resource?: string, limit: number = 50): Record<string, unknown>[] {
        let entries = this.auditLog;
        if (resource) entries = entries.filter(e => e.resource === resource);
        return entries.slice(-limit);
    }

    private validatePayment(amount: number, currency: string): void {
        if (!SUPPORTED_CURRENCIES.includes(currency)) {
            throw new ValidationError(`Unsupported currency: ${currency}`, "currency");
        }
        if (amount <= 0) throw new ValidationError("Amount must be positive", "amount");
        if (amount > 999999) throw new ValidationError("Amount exceeds maximum", "amount");
    }
}

File: benchmarks/fixtures/webapp_rb/services/payment/processor.rb
# Payment processor with caching + audit mixins.

require_relative '../../utils/helpers'
require_relative '../cacheable'
require_relative '../base'
require_relative '../../auth/service'
require_relative '../../database/connection'
require_relative '../../database/queries'
require_relative '../../events/dispatcher'
require_relative '../../exceptions'

SUPPORTED_CURRENCIES = %w[USD EUR GBP JPY CAD].freeze

# Payment processor with caching and audit trail.
class PaymentProcessor < BaseService
  include Cacheable
  include Auditable

  def initialize(db, events)
    super()
    @events = events
    @queries = PaymentQueries.new(db)
  end

  # Process a payment.
  def process_payment(user_id, amount, currency, method = 'card')
    _log("Processing payment: user=#{user_id}, amount=#{amount} #{currency}")
    validate_payment(amount, currency)
    txn_id = generate_request_id
    cache_key = "payment:#{user_id}:#{amount}:#{currency}"
    if cache_get(cache_key)
      raise PaymentError.new('Duplicate payment', transaction_id: txn_id)
    end
    begin
      @queries.create_payment(user_id, amount, currency, txn_id)
      @queries.update_status(txn_id, 'completed')
    rescue StandardError => e
      raise PaymentError.new("Payment failed: #{e}", transaction_id: txn_id)
    end
    cache_set(cache_key, txn_id, 300)
    record_audit('payment.processed', user_id, "payment:#{txn_id}", { amount: amount, currency: currency, method: method })
    @events.emit('payment.completed', { transaction_id: txn_id, user_id: user_id, amount: amount, currency: currency })
    { transaction_id: txn_id, status: 'completed', amount: amount, currency: currency }
  end

  # Refund a payment.
  def refund(transaction_id, reason = '')
    _log("Refunding: #{transaction_id}")
    payment = @queries.find_by_transaction_id(transaction_id)
    raise NotFoundError.new('Payment', transaction_id) unless payment

    @queries.update_status(transaction_id, 'refunded')
    record_audit('payment.refunded', 'system', "payment:#{transaction_id}", { reason: reason })
    @events.emit('payment.refunded', { transaction_id: transaction_id, reason: reason })
    { transaction_id: transaction_id, status: 'refunded' }
  end

  private

  def validate_payment(amount, currency)
    unless SUPPORTED_CURRENCIES.include?(currency)
      raise ValidationError.new("Unsupported currency: #{currency}", field: :currency)
    end
    raise ValidationError.new('Amount must be positive', field: :amount) if amount <= 0
    raise ValidationError.new('Amount exceeds maximum', field: :amount) if amount > 999_999
  end
end
